src/NUnitFramework/framework/Internal/Execution/ParallelWorkItemDispatcher.cs

```diff
@@ -244,41 +276,46 @@ private void RestoreQueues()
 
         private void OnEndOfShift(object sender, EventArgs ea)
         {
-            // This shift ended, so see if there is work in any other
-            // TODO: Temp fix - revisit later when object model is redesigned.
-            if (StartNextShift())
-                return;
+            ShiftFinished?.Invoke(sender as WorkShift);
```

Comment:
This one is production code. Can we direct-cast sender to `WorkShift` or at least throw an `ArgumentException`? Would seem to indicate abuse and safe casts create a secondary breach of expectations (raising ShiftFinished with a null WorkShift) while hiding the root cause.

Reply:
Changing delegate to be type-safe.